share/pot/export.sh

```diff
@@ -26,21 +27,63 @@ export-help() {
 _export_pot()
 {
 	# shellcheck disable=SC3043
-	local _pname _dset _snap _tag _dir _file _clvl
+	local _pname _dset _snap _tag _check_tag _prev_tag _prev_snap
+	# shellcheck disable=SC3043
+	local  _dir _file _clvl _meta _origin _origin_pname_snapshot
+	# shellcheck disable=SC3043
+	local _origin_pname _origin_snapshot _origin_tag _highest_version
 	_pname="$1"
 	_snap="$2"
 	_tag="$3"
-	_dir="$4"
-	_clvl="$5"
+	_check_tag="$4"
+	_dir="$5"
+	_clvl="$6"
 	_file="${_dir}/${_pname}_${_tag}.xz"
 	_dset="${POT_ZFS_ROOT}/jails/$_pname"
+        _meta="-"
+
+	_prev_tag=$(zfs get -H :pot.tag "${_dset}" | awk '{ print $3 }')
+	_prev_snap=$(zfs get -H :pot.snap "${_dset}" | awk '{ print $3 }')
+	if [ "$_check_tag" = "YES" ] && \
+	   [ -n "$_prev_tag" ] && [ "$_prev_tag" != "-" ]; then
+		if [ "$_prev_tag" = "$_tag" ] && [ "$_prev_snap" != "$_snap" ]; then
+			_error "Already exported a different snapshot tagged as this version"
+			exit 1
+		fi
+		_highest_version="$( \
+		  (echo "$_tag"; echo "$_prev_tag") | sort -V | tail -n1)"
+		if [ "$_tag" != "$_highest_version" ]; then
+			_error "Tag version lower than the previously exported one"
+			exit 1
+		fi
+	fi
+
+	_origin=$(zfs get -H origin "${_dset}/m" | awk '{ print $3 }')
+	if [ -n "$_origin" ] && [ "$_origin" != "-" ]; then
+		#_origin_dset=$(echo ${_origin} | awk -F@ '{ print $1 }')
+		_origin=$(echo "${_origin}" | sed 's|/m@|@|g')
+		_origin_pname_snapshot=$(basename "${_origin}")
+		_origin_pname=$(echo "${_origin_pname_snapshot}" | awk -F\@ '{ print $1 }')
+		_origin_snapshot=$(echo "${_origin_pname_snapshot}" | awk -F@ '{ print $2 }')
+		_origin_tag=$(zfs get -H :pot.tag "${_origin}" | awk '{ print $3 }')
+		if [ -z "$_origin_tag" ] || [ "$_origin_tag" = "-" ]; then
+			_error "Origin ${_origin_pname} has no :pot.tag, please export first"
+			return 1 # false
+		fi
+		_meta="${_origin_pname}:${_origin_tag}@${_origin_snapshot}"
+		#_origin=$(zfs get -H origin "${_origin_dset}" | awk '{ print $3 }')
+	fi
+
 	if ! zfs send -R "${_dset}"@"${_snap}" | xz -"${_clvl}" -T 0 > "${_file}" ; then
 		rm -f "${_file}"
 		return 1 # false
 	elif [ ! -r "${_file}" ]; then
-		return 1 # fasle
+		return 1 # false
 	else
-		skein1024 -q "${_file}" > "${_file}.skein"
+		echo "$_meta" > "${_file}.meta"
+		(cat "${_file}" "${_file}.meta") | skein1024 -q > "${_file}.skein"
+		zfs set :pot.tag="${_tag}" "${_dset}"
+		zfs set :pot.snap="${_snap}" "${_dset}"
```

Comment:
while `:pot.tag` is some sort of label attached to a specific snapshot and provides an additional piece of information, I don't understand the need of `:pot.snap`: it's the snapshot name

Reply:
_pot.snap_ is used earlier in the code (line 49):

```
if [ "$_prev_tag" = "$_tag" ] && [ "$_prev_snap" != "$_snap" ]; then
 			_error "Already exported a different snapshot tagged as this version"
fi
```

To allow having this check, it's important to have _pot.snap_ on the dataset.

The idea of this check is to allow exporting  the same tag multiple times, as long as the snapshot it's based on stays the same (and vice versa). Otherwise, layered pots based on that tag/snapshot will become inconsistent and can't be installed properly any more.

Even without layers I found that check to be quite useful to avoid creating the same version with different snapshots, like in:

    pot snapshot -p freebsd
    pot export -c -p freebsd -t 1.0 # that's okay
    pot snapshot -rp freebsd
    pot export -c -p freebsd -t 1.0 # that's NOT okay, we already exported 1.0 based on a different snapshot
    pot export -c -p freebsd -t 1.1 # that's okay again

This of course doesn't have historic information, but as long as you're passing `-c` to `pot export`, versions can't go backwards and you don't export different snapshots by accident.

As this happened to me many times while using pot (ending up with re-creating the same version of an image by accident, which caused issues with not doing updates later on) I added this check and found it to be quite useful.

Reply:
Find with me!